perf/top_address.go

```diff
@@ -0,0 +1,113 @@
+package perf
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/prque"
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var addrStatsEnabled, _ = getEnvBool("METRICS_TOP_ADDRESS_ENABLED")
+
+type AddrTxStats struct {
+	Map   map[string]int64 `json:"map"`
+	Start time.Time        `json:"start"`
+	End   time.Time        `json:"end"`
+}
+
+var eoaMap = make(map[string]int64)
+var contractMap = make(map[string]int64)
+
+var start = time.Now()
+
+func UpdateTopEOAStats(address common.Address) {
+	if !addrStatsEnabled {
+		return
+	}
+	inc(eoaMap, address.Hex())
+}
+
+func UpdateTopContractStats(address common.Address) {
+	if !addrStatsEnabled {
+		return
+	}
+	inc(contractMap, address.Hex())
+}
+
+func inc(m map[string]int64, address string) {
+	if val, ok := m[address]; ok {
+		m[address] = val + 1
+	} else {
+		m[address] = 1
+	}
+}
+
+func getTopEOA(topN int) AddrTxStats {
+	topM := top(eoaMap, topN)
+	return AddrTxStats{
+		Map:   topM,
+		Start: start,
+		End:   time.Now(),
+	}
+}
+
+func getTopContract(topN int) AddrTxStats {
+	topM := top(contractMap, topN)
+	return AddrTxStats{
+		Map:   topM,
+		Start: start,
+		End:   time.Now(),
+	}
+}
+
+func top(m map[string]int64, topN int) (topM map[string]int64) {
```

Comment:
I think there is a race condition on map `m`, since another routine may add item into map `m`, while this func is doing iterate over it.

Reply:
added lock